feat(#784): integrate bop_toolkit scripts into the bop writer

wboerdijk opened the pull request:
Suggests an implementation for #784. Compared to the original scripts (https://github.com/thodan/bop_toolkit/blob/master/scripts/calc_gt_masks.py, https://github.com/thodan/bop_toolkit/blob/master/scripts/calc_gt_info.py and https://github.com/thodan/bop_toolkit/blob/master/scripts/calc_gt_coco.py), this implementation only supports `vispy` renderer, `jpg` rgb images and `png` depth files. Visualization (e.g. https://github.com/thodan/bop_toolkit/blob/0a2cd868765ffb625c8263235ff071399a1c8007/scripts/calc_gt_info.py#L178) is removed.

The implementation should support the `chunk` formatted saving of the renderings, i.e. if there are existing chunk dirs where the first n renderings are appended to, and (multiple) new chunk dirs are created during writing for saving the remaining renderings, then gt masks, info and coco annotations are (re-)calculated for all folders where the bop writer has written to.

cornerfarmer commented:
BlenderProcHelper result for run 444:

BlenderProcHelper has started, waiting for first results.

cornerfarmer commented:
BlenderProcHelper result for run 445:

BlenderProcHelper has started, waiting for first results.

wboerdijk commented:
There's one more issue: installing the bop toolkit as [here](https://github.com/DLR-RM/BlenderProc/blob/de78d3e0c3ab2f2fd3803d89df352671702a823f/blenderproc/python/loader/BopLoader.py#L220-L241) doesnt install all required packages as they aren't listed in the [setup.py](https://github.com/thodan/bop_toolkit/blob/master/setup.py) but in a [requirements.txt](https://github.com/thodan/bop_toolkit/blob/master/requirements.txt). For now I installed all required packages separately, but we could also move this to the boploader if you prefer. Or, if there's any solution for installing from git including a `requirements.txt` (didn't find anything by googling), then I'm also open for that!

MartinSmeyer commented:
> There's one more issue: installing the bop toolkit as [here](https://github.com/DLR-RM/BlenderProc/blob/de78d3e0c3ab2f2fd3803d89df352671702a823f/blenderproc/python/loader/BopLoader.py#L220-L241) doesnt install all required packages as they aren't listed in the [setup.py](https://github.com/thodan/bop_toolkit/blob/master/setup.py) but in a [requirements.txt](https://github.com/thodan/bop_toolkit/blob/master/requirements.txt). For now I installed all required packages separately (https://github.com/wboerdijk/BlenderProc/blob/b16c5a9ffaaa1053aa85b9771d8d296fe47aed16/blenderproc/python/writer/BopWriterUtility.py#L107), but we could also move this to the boploader if you prefer. Or, if there's any solution for installing from git including a `requirements.txt` (didn't find anything by googling), then I'm also open for that!

If we only need vispy and PyOpenGL, I could also add those to the setup.py in the bop_toolkit. 

I think it's good to setup the bop_toolkit in the BopWriter like you do currently, because the BopLoader is not always used.

MartinSmeyer commented on blenderproc/python/writer/BopWriterUtility.py:
```diff
@@ -18,13 +19,15 @@
               depth_scale: float = 1.0, jpg_quality: int = 95, save_world2cam: bool = True,
-              ignore_dist_thres: float = 100., m2mm: bool = True, frames_per_chunk: int = 1000):
+              ignore_dist_thres: float = 100., m2mm: bool = True, frames_per_chunk: int = 1000,
+              calc_mask_info_coco: bool = False, delta: int = 15):
```
I think we can even set the default to True
```
calc_mask_info_coco: bool = True
```

MartinSmeyer commented on blenderproc/python/writer/BopWriterUtility.py:
```diff
@@ -82,13 +87,56 @@ def write_bop(output_dir: str, target_objects: Optional[List[MeshObject]] = None
                            f"Either remove the dataset parameter or assign custom property 'bop_dataset_name'"
                            f" to selected objects")
 
+    if calc_mask_info_coco:
+        # It might be that a chunk dir already exists where the writer appends frames.
+        # If one (or multiple) more chunk dirs are created to save the rendered frames to,
+        # mask/info/coco annotations need to be calculated for all of them
+        chunk_dirs = sorted(glob.glob(os.path.join(chunks_dir, '*')))
+        chunk_dirs = [d for d in chunk_dirs if os.path.isdir(d)]
+        starting_chunk_dir = sorted(chunk_dirs)[-1] if chunk_dirs else None
```
can we move this part down to l105+ ?

MartinSmeyer replied:
ah now I get the logic, you check whether a new chunk_dir was created and only then run the calc_gt_info_mask stuff for the last chunk_dir.

wboerdijk replied:
hm I don't think so; I'm trying to determine the last chunk dir the writer will be appending to. Let's say you write 25 files per chunk, and in a previous run one only has written 20 frames. gt, info and coco are then only calculated for these 20 frames. if you run the script again and want to render e.g. 40 frames, the writer fills the five remaining frames from the first chunk dir, and creates two additional ones (5 + 25 + 10). before writing we need to determine the currently last chunk dir to be able to also calculate gt info and coco for the 5 additional frames, and all the upcoming chunk dirs.

MartinSmeyer replied:
Ah okay, thanks for the explanation. But one chunk is usually 1000 images, are you sure you don't repeat the calculation of the masks/info/coco for the same images? In the BOP Challenge examples the BopWriter is called every 25 frames, i.e. after each scene (!= chunk). 

I don't see where you update the image ids, to me it looks like you compute it for all images in `scene_gt.json` every time, but I could be wrong:

https://github.com/DLR-RM/BlenderProc/blob/5c4e0f1d22aad92ff2e64bf87995c8256bdf9481/blenderproc/python/writer/BopWriterUtility.py#L516-L517

cornerfarmer commented:
BlenderProcHelper result for run 453:

BlenderProcHelper has started, waiting for first results.

wboerdijk commented:
> If we only need vispy and PyOpenGL, I could also add those to the setup.py in the bop_toolkit.

Yes, that would be great! Then I'm moving back to using `_BopLoader.setup_bop_toolkit()`:

https://github.com/wboerdijk/BlenderProc/blob/5c4e0f1d22aad92ff2e64bf87995c8256bdf9481/blenderproc/python/writer/BopWriterUtility.py#L107

MartinSmeyer commented:
> > If we only need vispy and PyOpenGL, I could also add those to the setup.py in the bop_toolkit.
> 
> Yes, that would be great! Then I'm moving back to using `_BopLoader.setup_bop_toolkit()`:
> 
> https://github.com/wboerdijk/BlenderProc/blob/5c4e0f1d22aad92ff2e64bf87995c8256bdf9481/blenderproc/python/writer/BopWriterUtility.py#L107

Updated the `bop_toolkit`. But the `bop_toolkit` is not automatically reinstalled if it already exists (and shouldn't be if we call it every time), so maybe we should still add the two dependencies again after this line to make sure?

https://github.com/wboerdijk/BlenderProc/blob/5c4e0f1d22aad92ff2e64bf87995c8256bdf9481/blenderproc/python/loader/BopLoader.py#L238

MartinSmeyer commented:
Do we need the full `setup_bop_toolkit()`, it prints some stuff and set env varibales. Maybe this is sufficient:
```python
SetupUtility.setup_pip(["git+https://github.com/thodan/bop_toolkit"])
```

cornerfarmer commented:
BlenderProcHelper result for run 459:

BlenderProcHelper has started, waiting for first results.

wboerdijk commented:
I reverted that part, and added the two dependencies.
I also fixed a bug due to the renderer being a singleton instance - to correctly render we initialize the instance again with new height/width since any form of deconstruction doesn't work here:
https://github.com/wboerdijk/BlenderProc/blob/3ddc273ab1cf3e7a2efc1b9c723830274232703c/blenderproc/python/writer/BopWriterUtility.py#L475
I compared the output visually and with the default scripts from the bop toolkit, and everything looks fine to me. Nevertheless, is there any insanity check I can run?

MartinSmeyer commented:
> I reverted that part, and added the two dependencies.
I also fixed a bug due to the renderer being a singleton instance - to correctly render we initialize the instance again with new height/width since any form of deconstruction doesn't work here:
https://github.com/wboerdijk/BlenderProc/blob/3ddc273ab1cf3e7a2efc1b9c723830274232703c/blenderproc/python/writer/BopWriterUtility.py#L475
I compared the output visually and with the default scripts from the bop toolkit, and everything looks fine to me. Nevertheless, is there any insanity check I can run?

Thanks, I guess you could run the bop_challenge examples with 2-3 scenes, they are not part of the BlenderProcHelper.

MartinSmeyer commented:
My question is if you repeat the computation of masks/info/coco everytime the write_bop() is called for all images in a chunk or just the new ones.

cornerfarmer commented:
BlenderProcHelper result for run 462:

BlenderProcHelper has started, waiting for first results.

wboerdijk commented:
> My question is if you repeat the computation of masks/info/coco everytime the write_bop() is called for all images in a chunk or just the new ones.

Yes, you are absolutely right. I fixed this in the latest commits; now the gt_info_coco is calculated only for newly saved frames.
During testing on the bop challenge examples I also fixed a bug; now gt_info_coco is calculated for all dataset objects, even if objects are hidden.
I created a few sample scenes for every bop_challenge dataset, and visually inspected the outputs; they look good to me.

MartinSmeyer commented on blenderproc/python/writer/BopWriterUtility.py:
```diff
@@ -425,3 +463,373 @@ def write_frames(chunks_dir: str, dataset_objects: list, depths: Optional[List[n
+        from bop_toolkit_lib import inout, misc, visibility, renderer
+        # pylint: enable=import-outside-toplevel
+
+        width = bpy.context.scene.render.resolution_x
+        height = bpy.context.scene.render.resolution_y
+        # the renderer instance is a singleton - if we call this function twice, we obtain the same instance again here
+        ren = renderer.create_renderer(width=width, height=height, renderer_type='vispy', mode='depth')
+        # hence, we need to call the `__init__` again to properly set up the rendering
+        # pylint: disable=unnecessary-dunder-call
+        ren.__init__(width=width, height=height)
+        # pylint: enable=unnecessary-dunder-call
+
+        # add objects
+        # for numpy>=1.20, np.float is deprecated: https://numpy.org/doc/stable/release/1.20.0-notes.html#deprecations
+        np.float = float
+        dataset_objects = []
+        if dataset != "":
+            for obj in get_all_mesh_objects():
+                if "bop_dataset_name" in obj.blender_obj:
+                    # opposite to above, we write annotations also for hidden objects
+                    if obj.blender_obj["bop_dataset_name"] == dataset:
+                        dataset_objects.append(obj)
+        else:
+            # if no dataset is defined, write annotations for all objects
+            dataset_objects = get_all_mesh_objects()
+
+        for obj in dataset_objects:
+            ren.add_object(obj_id=obj.get_cp('category_id'), model_path=obj.get_cp('model_path'))
+
+        # Load scene info and ground-truth poses.
+        chunk_dirs = sorted(glob.glob(os.path.join(chunks_dir, '*')))
+        chunk_dirs = [d for d in chunk_dirs if os.path.isdir(d)]
+        chunk_dir_ids = [d.split('/')[-1] for d in chunk_dirs]
+        chunk_dirs = chunk_dirs[chunk_dir_ids.index(f"{starting_chunk_id:06d}"):]
+
+        for dir_counter, chunk_dir in enumerate(chunk_dirs):
+            last_chunk_gt_fpath = os.path.join(chunk_dir, 'scene_gt.json')
+            last_chunk_camera_fpath = os.path.join(chunk_dir, 'scene_camera.json')
+            scene_gt = _BopWriterUtility.load_json(last_chunk_gt_fpath, keys_to_int=True)
+            scene_camera = _BopWriterUtility.load_json(last_chunk_camera_fpath, keys_to_int=True)
```
Since this is shared among `calc_gt_info()` and `calc_gt_mask()` and partly `calc_gt_coco()` can we move it in front? It should also be more efficient.

wboerdijk replied:
Yes, I moved them outside! The only thing that needs to stay inside is the renderer init since the image dims are different.

cornerfarmer commented:
BlenderProcHelper result for run 466:

BlenderProcHelper has started, waiting for first results.
